src/Npgsql/NpgsqlParameter.cs

```diff
@@ -506,60 +495,254 @@ public sealed override string SourceColumn
+        // This step isn't part of BindValue because we need to know the PgTypeId beforehand for things like SchemaOnly with null values.
+        // We never reuse resolutions for resolvers across executions as a mutable value itself may influence the result.
+        // TODO we could expose a property on a Converter/TypeInfo to indicate whether it's immutable, at that point we can reuse.
+        if (!previouslyBound || TypeInfo is PgResolverTypeInfo)
         {
-            var parameterName = !string.IsNullOrEmpty(ParameterName) ? ParameterName : $"${Collection?.IndexOf(this) + 1}";
-            ThrowHelper.ThrowInvalidOperationException($"Parameter '{parameterName}' must have either its NpgsqlDbType or its DataTypeName or its Value set");
+            ResetConverterResolution();
+            var resolution = ResolveConverter(TypeInfo!);
+            Converter = resolution.Converter;
+            PgTypeId = resolution.PgTypeId;
+        }
+
+        bool TryGetRepresentationalTypeId(string dataTypeName, out PgTypeId pgTypeId)
+        {
+            if (options.TypeCatalog.TryGetPostgresTypeByName(dataTypeName, out var pgType))
+            {
+                pgTypeId = options.ToCanonicalTypeId(pgType.GetRepresentationalType());
+                return true;
+            }
+
+            pgTypeId = default;
+            return false;
         }
     }
 
-    internal void Bind(TypeMapper typeMapper)
+    private protected virtual PgConverterResolution ResolveConverter(PgTypeInfo typeInfo) => typeInfo.GetObjectResolution(_value);
+
+    /// Bind the current value to the type info, truncate (if applicable), take its size, and do any final validation before writing.
+    internal void Bind(out DataFormat format, out Size size)
     {
-        ResolveHandler(typeMapper);
-        FormatCode = Handler!.PreferTextWrite ? FormatCode.Text : FormatCode.Binary;
+        if (TypeInfo is null)
+            ThrowHelper.ThrowInvalidOperationException("Missing type info.");
+
+        if (!TypeInfo.SupportsWriting)
+            ThrowHelper.ThrowNotSupportedException($"Cannot write values for parameters of type '{TypeInfo.Type}' and postgres type '{TypeInfo.Options.TypeCatalog.GetDataTypeName(PgTypeId).DisplayName}'.");
+
+        // We might call this twice, once during validation and once during WriteBind, only compute things once.
+        if (WriteSize is not null)
+        {
+            format = Format;
+            size = WriteSize.Value;
+            return;
+        }
+
+        // Handle Size truncate behavior for a predetermined set of types and pg types.
+        // Doesn't matter if we 'box' value, all supported types are reference types.
+        if (_size > 0 && Converter!.TypeToConvert is var type &&
+            (type == typeof(string) || type == typeof(char[]) || type == typeof(byte[]) || type == typeof(Stream)) &&
+            Value is { } value)
+        {
+            var dataTypeName = TypeInfo!.Options.GetDataTypeName(PgTypeId);
```

Comment:
nit: switch for all this?

Reply:
It needs equals so that'll just be `when` arms